pkg/server/sessions/session.go

```diff
@@ -0,0 +1,220 @@
+	now := time.Now()
+	return &Session{
+		id:               sessionID,
+		state:            ACTIVE,
+		user:             user,
+		databaseID:       databaseID,
+		creationTime:     now,
+		lastActivityTime: now,
+		lastHeartBeat:    now,
+		transactions:     make(map[string]transactions.Transaction),
+		log:              log,
+	}
+}
+
+func (s *Session) NewTransaction(sqlTx *sql.SQLTx, mode schema.TxMode, db database.DB) (transactions.Transaction, error) {
+	s.Lock()
+	defer s.Unlock()
+	if mode == schema.TxMode_READ_WRITE {
+		if s.readWriteTxOngoing {
+			return nil, ErrReadWriteTxOngoing
+		}
+		s.readWriteTxOngoing = true
+	}
+	transactionID := xid.New().String()
+	tx := transactions.NewTransaction(sqlTx, transactionID, mode, db, s.id)
+	s.transactions[transactionID] = tx
+	return tx, nil
+}
+
+func (s *Session) RemoveTransaction(transactionID string) {
+	s.Lock()
+	defer s.Unlock()
+	if _, ok := s.transactions[transactionID]; ok {
+		s.removeTransaction(transactionID)
+	}
+}
+
+// not thread safe
+func (s *Session) removeTransaction(transactionID string) {
+	if tx, ok := s.transactions[transactionID]; ok {
+		if tx.GetMode() == schema.TxMode_READ_WRITE {
+			s.readWriteTxOngoing = false
+		}
+		delete(s.transactions, transactionID)
+	}
+}
+
+func (s *Session) RollbackTransactions() {
+	s.Lock()
+	defer s.Unlock()
+	for _, tx := range s.transactions {
+		s.log.Debugf("Deleting transaction %s", tx.GetID())
+		err := tx.Rollback()
+		s.log.Errorf("transaction %s cancelled with error %v", tx.GetID(), err)
+		s.removeTransaction(tx.GetID())
+	}
+}
+
+func (s *Session) GetTransaction(transactionID string) (transactions.Transaction, error) {
+	s.RLock()
+	defer s.RUnlock()
+	tx, ok := s.transactions[transactionID]
+	if !ok {
+		return nil, transactions.ErrTransactionNotFound
+	}
+	return tx, nil
+}
+
+func GetSessionIDFromContext(ctx context.Context) (string, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", ErrNoSessionAuthDataProvided
+	}
+	authHeader, ok := md["sessionid"]
+	if !ok || len(authHeader) < 1 {
+		return "", ErrNoSessionAuthDataProvided
+	}
+	sessionID := authHeader[0]
+	if sessionID == "" {
+		return "", ErrNoSessionIDPresent
+	}
+	return sessionID, nil
+}
+
+func GetTransactionIDFromContext(ctx context.Context) (string, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", ErrNoTransactionAuthDataProvided
+	}
+	authHeader, ok := md["transactionid"]
+	if !ok || len(authHeader) < 1 {
+		return "", ErrNoTransactionAuthDataProvided
+	}
+	sessionID := authHeader[0]
+	if sessionID == "" {
+		return "", ErrNoTransactionIDPresent
+	}
+	return sessionID, nil
+}
+
+func (s *Session) GetUser() *auth.User {
+	s.RLock()
+	defer s.RUnlock()
+	return s.user
+}
+
+func (s *Session) GetDatabaseID() int64 {
+	s.RLock()
+	defer s.RUnlock()
+	return s.databaseID
+}
+
+func (s *Session) SetDatabaseID(databaseID int64) {
```

Comment:
it may only be set when session is opened

Reply:
It's used also when changing database with UseDatabase

Reply:
wondering if it's worth changing database in the session... as long as it doesn't affect ongoing transaction it may be ok, but one possibility could be to just allow selecting the database when creating a session and simplifying the api by deprecating useDatabase...

Reply:
Transaction is bind to a single database and is not exposed anything to change on it